Compilation issue in MacOS Silicon: This header is only meant to be used on x86 and x64 architecture

I am trying to compile the MVE in MacOS (Sonoma 14.4.1), with CPU M1 Max.

1. I installed libraries with brew command:
```
arch -arm64 brew install libjpeg libtiff
```
2. Because it was still complaining about missing headers, I modified Makefile as following:
```diff
diff --git a/libs/mve/Makefile b/libs/mve/Makefile
index 6240dda..853ecc4 100644
--- a/libs/mve/Makefile
+++ b/libs/mve/Makefile
@@ -3,8 +3,8 @@ TARGET := libmve.a
 include ${MVE_ROOT}/Makefile.inc

 # Position independent code (-fPIC) is required for the UMVE plugin system.
-CXXFLAGS += -fPIC -I${MVE_ROOT}/libs
-LDLIBS += -lpng -ltiff -ljpeg
+CXXFLAGS += -fPIC -I${MVE_ROOT}/libs -I/opt/homebrew/include
+LDLIBS += -lpng -ltiff -ljpeg -L/opt/homebrew/lib
```
3. I executed `make -j8`. Now I have at least 20 compilation errors "This header is only meant to be used on x86 and x64 architecture", first lines with error:
```
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C sfm
c++ -Wall -Wextra -Wundef -pedantic -march=native -funsafe-math-optimizations -fno-math-errno -std=c++11 -g -O3 -pthread -I../../libs   -c -MM ba_linear_solver.cc bundle_adjustment.cc bundler_common.cc bundler_features.cc bundler_incremental.cc bundler_init_pair.cc bundler_intrinsics.cc bundler_matching.cc bundler_tracks.cc camera_database.cc cascade_hashing.cc exhaustive_matching.cc extract_focal_length.cc feature_set.cc fundamental.cc homography.cc matching.cc nearest_neighbor.cc pose_p3p.cc ransac.cc ransac_fundamental.cc ransac_homography.cc ransac_pose_p3p.cc sift.cc surf.cc triangulate.cc visualizer.cc >Makefile.dep
In file included from nearest_neighbor.cc:42:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include/emmintrin.h:14:2: error: "This header is only meant to be used on x86 and x64 architecture"
#error "This header is only meant to be used on x86 and x64 architecture"
 ^
In file included from nearest_neighbor.cc:42:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include/emmintrin.h:17:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include/xmmintrin.h:14:2: error: "This header is only meant to be used on x86 and x64 architecture"
#error "This header is only meant to be used on x86 and x64 architecture"
 ^
In file included from nearest_neighbor.cc:42:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include/emmintrin.h:17:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include/xmmintrin.h:17:
```

Do anybody tried to compile for MacOS Silicon?

@andre-schulz thank you, it works. Now it fails on linking stage:
```
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C apps
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C bundle2pset
c++  bundle2pset.o ../../libs/mve/libmve.a ../../libs/util/libmve_util.a -lpng -ltiff -ljpeg  -L/opt/homebrew/lib -o bundle2pset
Undefined symbols for architecture arm64:
  "mve::geom::save_ply_mesh(std::__1::shared_ptr<mve::TriangleMesh const>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, mve::geom::SavePLYOptions const&)", referenced from:
      _main in bundle2pset.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bundle2pset] Error 1
make[1]: *** [all] Error 2
make: *** [all] Error 2
```

@andre-schulz 

The original full output (`make -j8`):
```sh
% make -j8
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C libs
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C math
#ar rcs libmve_math.a
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C util
make[2]: `libmve_util.a' is up to date.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C mve
make[2]: `libmve.a' is up to date.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C sfm
make[2]: `libmve_sfm.a' is up to date.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C dmrecon
make[2]: `libmve_dmrecon.a' is up to date.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C fssr
make[2]: `libmve_fssr.a' is up to date.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ogl
make[2]: `libmve_ogl.a' is up to date.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C apps
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C bundle2pset
c++  bundle2pset.o ../../libs/mve/libmve.a ../../libs/util/libmve_util.a -lpng -ltiff -ljpeg  -L/opt/homebrew/lib -o bundle2pset
Undefined symbols for architecture arm64:
  "mve::geom::save_ply_mesh(std::__1::shared_ptr<mve::TriangleMesh const>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, mve::geom::SavePLYOptions const&)", referenced from:
      _main in bundle2pset.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bundle2pset] Error 1
make[1]: *** [all] Error 2
make: *** [all] Error 2
```

Then I repeated the problematic command with `-v` in the problematic folder:
```sh
% c++ -v bundle2pset.o ../../libs/mve/libmve.a ../../libs/util/libmve_util.a -lpng -ltiff -ljpeg  -L/opt/homebrew/lib -o bundle2pset
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -arch arm64 -platform_version macos 14.0.0 14.4 -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -o bundle2pset -L/opt/homebrew/lib -L/usr/local/lib bundle2pset.o ../../libs/mve/libmve.a ../../libs/util/libmve_util.a -lpng -ltiff -ljpeg -lc++ -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture arm64:
  "mve::geom::save_ply_mesh(std::__1::shared_ptr<mve::TriangleMesh const>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, mve::geom::SavePLYOptions const&)", referenced from:
      _main in bundle2pset.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

ObjDump shows that `bundle2pset.o` is arm64:
```sh
% objdump -a bundle2pset.o
bundle2pset.o:	file format mach-o arm64
```

--- a/libs/sfm/nearest_neighbor.cc
+++ b/libs/sfm/nearest_neighbor.cc
@@ -39,8 +39,12 @@
 
 #include <algorithm>
 #include <iostream>
-#include <emmintrin.h> // SSE2
-#include <pmmintrin.h> // SSE3
+#if defined(__SSE2__)
+#   include <emmintrin.h> // SSE2
+#endif
+#if defined(__SSE3__)
+#   include <pmmintrin.h> // SSE3
+#endif
 
 #include "sfm/nearest_neighbor.h"
 

